Tools/pyco/pyco.py

```diff
@@ -453,44 +458,70 @@ class HasValue(Protocol):
 
 class Builder:
     def __init__(self):
-        self.codeobjs: list[CodeObject] = []
-        self.strings: list[String | Redirect] = []
-        self.blobs: list[BlobConstant] = []
-        self.constants: list[ComplexConstant] = []
+        SLOW_VERSION = False
+        if SLOW_VERSION:
+            self.codeobjs: list[CodeObject] = []
+            self.strings: list[String | Redirect] = []
+            self.blobs: list[BlobConstant] = []
+            self.constants: list[ComplexConstant] = []
+        else:
+            self.codeobjs: Mapping[CodeObject | Redirect, Tuple(int, int)] = OrderedDict()
+            self.strings: Mapping[String | Redirect, Tuple(int, int)] = OrderedDict()
+            self.blobs: Mapping[BlobConstant | Redirect, Tuple(int, int)] = OrderedDict()
+            self.constants: Mapping[ComplexConstant | Redirect, Tuple(int, int)] = OrderedDict()
         self.locked = False
-
         self.co_strings_start = -1
 
     def lock(self):
         self.locked = True
 
-    def add(self, where: list[T], thing: T) -> int:
+    def add(self, where: list[T|Redirect], thing: T|Redirect) -> int:
         # TODO: Avoid O(N**2) lookup behavior
         # Look for a match
-        for index, it in enumerate(where):
-            if type(it.value) is type(thing.value) and it.value == thing.value:
+        if isinstance(where, Mapping):
+            if thing in where:
+                index, _ = where[thing]
                 return index
+        else:
+            for index, it in enumerate(where):
+                if type(it.value) is type(thing.value) and it.value == thing.value:
+                    return index
         # Append a new one
         index = len(where)
         assert not self.locked
-        where.append(thing)
+        if isinstance(where, Mapping):
+            where[thing] = [index, index]
+        else:
+            where.append(thing)
         return index
 
-    def add_redirect(self, where: list[T], target: int):
-        for index in range(self.co_strings_start, len(where)):
-            # Do we have one for this co?
-            it = self.strings[index]
-            if isinstance(it, Redirect) and it.target == target:
-                return index
-        return self.add(where, Redirect(target))
+    def add_redirect(self, where: list[T], thing: T, target: int):
+        if isinstance(where, Mapping):
+            thing_index, redirect_index = where[thing]
+            assert thing_index == target
+            if redirect_index >= self.co_strings_start:
+                # we already have a redirect for Thing in this co
+                # I'm not sure this can actually happen (co_names are unique, right?)
+                return redirect_index
+            index = self.add(where, Redirect(target))
+            where[thing][1] = index
```

Comment:
`where[thing] = (where[thing][0], index)`

Reply:
yeah that gave the mutating exceptions.